### PR TITLE
Add stream metrics support

### PR DESCRIPTION
**Pull request by Dean-Coakley**
### Proposed changes
Closes: https://github.com/nginxinc/nginx-prometheus-exporter/pull/1

* Add Stream upstream metrics
* Add Stream Server Zone metrics
* Add Stream Server metrics
* Update documentation to include stream documentation

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/nginx-prometheus-exporter/blob/master/CONTRIBUTING.md) guide
- [x] I have proven my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have ensured the README is up to date
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch on my own fork

### CODE REVIEW

File: collector/nginx_plus.go [hunk 1]
```diff
@@ -42,10 +48,24 @@ func NewNginxPlusCollector(nginxClient *plusclient.NginxClient, namespace string
 			"received":      newServerZoneMetric(namespace, "received", "Bytes received from clients", nil),
 			"sent":          newServerZoneMetric(namespace, "sent", "Bytes sent to clients", nil),
 		},
+		streamServerZoneMetrics: map[string]*prometheus.Desc{
+			"processing":     newStreamServerZoneMetric(namespace, "processing", "Client connections that are currently being processed", nil),
+			"connections":    newStreamServerZoneMetric(namespace, "connections", "Total connections", nil),
+			"sessions_2xx":   newStreamServerZoneMetric(namespace, "sessions", "Total sessions completed", prometheus.Labels{"code": "2xx"}),
+			"sessions_4xx":   newStreamServerZoneMetric(namespace, "sessions", "Total sessions completed", prometheus.Labels{"code": "4xx"}),
+			"sessions_5xx":   newStreamServerZoneMetric(namespace, "sessions", "Total sessions completed", prometheus.Labels{"code": "5xx"}),
+			"sessions_total": newStreamServerZoneMetric(namespace, "sessions", "Total sessions completed", prometheus.Labels{"code": "total"}),
```
Comment (pleshakov) [inline collector/nginx_plus.go]: not required. the total number can be obtained in Prometheus by aggregating 2xx,4xx,5xx
Also,  see responses metrics for http zones

File: collector/nginx_plus.go [hunk 2]
```diff
@@ -65,6 +85,22 @@ func NewNginxPlusCollector(nginxClient *plusclient.NginxClient, namespace string
 			"health_checks_fails":     newUpstreamServerMetric(namespace, "health_checks_fails", "Failed health checks", nil),
 			"health_checks_unhealthy": newUpstreamServerMetric(namespace, "health_checks_unhealthy", "How many times the server became unhealthy (state 'unhealthy')", nil),
 		},
+		streamUpstreamServerMetrics: map[string]*prometheus.Desc{
+			"state":                   newStreamUpstreamServerMetric(namespace, "state", "Current state"),
+			"active":                  newStreamUpstreamServerMetric(namespace, "active", "Active connections"),
+			"sent":                    newStreamUpstreamServerMetric(namespace, "sent", "Bytes sent to this server"),
+			"received":                newStreamUpstreamServerMetric(namespace, "received", "Bytes received to this server"),
```
Comment (pleshakov) [inline collector/nginx_plus.go]: -> Bytes received from this server.

File: collector/nginx_plus.go [hunk 3]
```diff
@@ -65,6 +85,22 @@ func NewNginxPlusCollector(nginxClient *plusclient.NginxClient, namespace string
...
+			"fails":                   newStreamUpstreamServerMetric(namespace, "fails", "Number of unsuccessful attempts to communicate with the server"),
+			"unavail":                 newStreamUpstreamServerMetric(namespace, "unavail", "How many times the server became unavailable for client connections (state 'unavail') due to the number of unsuccessful attempts reaching the max_fails threshold"),
+			"connections":             newStreamUpstreamServerMetric(namespace, "connections", "Total number of client connections forwarded to this server"),
+			"connect_time":            newStreamUpstreamServerMetric(namespace, "connect_time", "Average time to connect to the upstream server"),
+			"first_byte_time":         newStreamUpstreamServerMetric(namespace, "first_byte_time", "The average time to receive the first byte of data"),
```
Comment (pleshakov) [inline collector/nginx_plus.go]: -> Average time to receive the first byte of data

File: collector/nginx_plus.go [hunk 4]
```diff
@@ -65,6 +85,22 @@ func NewNginxPlusCollector(nginxClient *plusclient.NginxClient, namespace string
...
+			"response_time":           newStreamUpstreamServerMetric(namespace, "response_time", "Average time to get the full response from the server"),
```
Comment (pleshakov) [inline collector/nginx_plus.go]: -> Average time to receive the last byte of data

File: collector/nginx_plus.go [hunk 5]
```diff
@@ -65,6 +85,22 @@ func NewNginxPlusCollector(nginxClient *plusclient.NginxClient, namespace string
...
+			"health_checks_checks":    newStreamUpstreamServerMetric(namespace, "health_checks_checks", "Total health check requests"),
+			"health_checks_fails":     newStreamUpstreamServerMetric(namespace, "health_checks_fails", "Failed health checks"),
+			"health_checks_unhealthy": newStreamUpstreamServerMetric(namespace, "health_checks_unhealthy", "How many times the server became unhealthy (state 'unhealthy')"),
+			"downtime":                newStreamUpstreamServerMetric(namespace, "downtime", "Total time the server was in the 'unavail', 'checking', and 'unhealthy' states"),
```
Comment (pleshakov) [inline collector/nginx_plus.go]: Please remove downtime metric

File: collector/nginx_plus.go [hunk 6]
```diff
@@ -183,6 +249,43 @@ func (c *NginxPlusCollector) Collect(ch chan<- prometheus.Metric) {
...
 	}
+
+	for name, upstream := range stats.StreamUpstreams {
+		for _, peer := range upstream.Peers {
+			ch <- prometheus.MustNewConstMetric(c.streamUpstreamServerMetrics["state"],
+				prometheus.GaugeValue, upstreamServerStates[peer.State], name, peer.Server)
+			ch <- prometheus.MustNewConstMetric(c.streamUpstreamServerMetrics["active"],
+				prometheus.GaugeValue, float64(peer.Active), name, peer.Server)
+			ch <- prometheus.MustNewConstMetric(c.streamUpstreamServerMetrics["connections"],
+				prometheus.GaugeValue, float64(peer.Connections), name, peer.Server)
```
Comment (pleshakov) [inline collector/nginx_plus.go]: Counter

File: README.md [hunk 7]
```diff
@@ -70,7 +70,10 @@ Usage of ./nginx-prometheus-exporter:
     * [HTTP](http://nginx.org/en/docs/http/ngx_http_api_module.html#http_).
     * [SSL](http://nginx.org/en/docs/http/ngx_http_api_module.html#def_nginx_ssl_object).
     * [HTTP Server Zones](http://nginx.org/en/docs/http/ngx_http_api_module.html#def_nginx_http_server_zone).
+    * [Stream Server Zones](http://nginx.org/en/docs/http/ngx_http_api_module.html#def_nginx_stream_server_zone).
     * [HTTP Upsteams](http://nginx.org/en/docs/http/ngx_http_api_module.html#def_nginx_http_upstream). Note: for the `state` metric, the string values are converted to float64 using the following rule: `"up"` -> `1.0`, `"draining"` -> `2.0`, `"down"` -> `3.0`, `"unavail"` –> `4.0`, `"checking"` –> `5.0`, `"unhealthy"` -> `6.0`.
+    * [Stream Upsteams](http://nginx.org/en/docs/http/ngx_http_api_module.html#def_nginx_stream_upstream). Note: for the `state` metric, the string values are converted to float64 using the following rule: `"up"` -> `1.0`, `"down"` -> `3.0`, `"unavail"` –> `4.0`, `"checking"` –> `5.0`, `"unhealthy"` -> `6.0`.
```
Comment (pleshakov) [inline README.md]: Upsteams -> Upstreams
please fix for http as well

File: collector/nginx_plus.go (hunk 5 above)
Comment (Dean-Coakley) [reply collector/nginx_plus.go]: @pleshakov https://github.com/nginxinc/nginx-prometheus-exporter/blob/5e42f9a26e714d24d69387d3ffa2a920d9830f5d/collector/nginx_plus.go#L283 too?

Comment (pleshakov) [reply collector/nginx_plus.go]: yes

**Pull request merged**